File: src/modules/platform/input.c
#include <stdio.h>
#include <string.h>
#include "raylib.h"
#include "core/collections/sthashmap.h"
#include "input.h"

/// key is a string, value is a stlist of stkeys
StHashMap* __st_keymaps__;

void __st_init_input__()
{
    __st_keymaps__ = StHashMap_new(0);
}

void __st_free_input__()
{
    // sometimes you start losing your fucking mind
    StList* items = StHashMap_items(__st_keymaps__);
    for (nint i = 0; i < items->length; i++) {
        StTuple2* kvp = (StTuple2*)items->items[i];
        // we don't free the items because they're numbers
        StList_free(kvp->item2);
    }
    StList_free_with_items(items);

    StHashMap_free(__st_keymaps__);
}

bool StInput_is_key_just_pressed(StKey key)
{
    return IsKeyPressed(key);
}

bool StInput_is_key_just_released(StKey key)
{
    return IsKeyReleased(key);
}

bool StInput_is_key_held(StKey key)
{
    return IsKeyDown(key);
}

bool StInput_is_key_not_pressed(StKey key)
{
    return IsKeyUp(key);
}

bool StInput_is_mouse_button_just_pressed(StMouseButton btn)
{
    return IsMouseButtonPressed(btn);
}

bool StInput_is_mouse_button_just_released(StMouseButton btn)
{
    return IsMouseButtonReleased(btn);
}

bool StInput_is_mouse_button_held(StMouseButton btn)
{
    return IsMouseButtonDown(btn);
}

bool StInput_is_mouse_button_not_pressed(StMouseButton btn)
{
    return IsMouseButtonUp(btn);
}

void StInput_add_keymap(const char* keymap, StKey key)
{
    StList* leliste;
    if (!StHashMap_contains(__st_keymaps__, keymap)) {
        leliste = StList_new(1);
        StHashMap_set(__st_keymaps__, keymap, leliste);
    }
    else {
        leliste = StHashMap_get(__st_keymaps__, keymap);
    }

    StList_add(leliste, (void*)key);
}

bool StInput_is_keymap_just_pressed(const char* keymap)
{
    if (!StHashMap_contains(__st_keymaps__, keymap)) return false;

    StList* keys = StHashMap_get(__st_keymaps__, keymap);
    for (nint i = 0; i < keys->length; i++) {
        if (StInput_is_key_just_pressed((StKey)keys->items[i])) return true;
    }
    return false;
}

bool StInput_is_keymap_just_released(const char* keymap)
{
    if (!StHashMap_contains(__st_keymaps__, keymap)) return false;

    StList* keys = StHashMap_get(__st_keymaps__, keymap);
    for (nint i = 0; i < keys->length; i++) {
        if (StInput_is_key_just_released((StKey)keys->items[i])) return true;
    }
    return false;
}

bool StInput_is_keymap_held(const char* keymap)
{
    if (!StHashMap_contains(__st_keymaps__, keymap)) return false;

    StList* keys = StHashMap_get(__st_keymaps__, keymap);
    for (nint i = 0; i < keys->length; i++) {
        if (StInput_is_key_held((StKey)keys->items[i])) return true;
    }
    return false;
}

bool StInput_is_keymap_not_pressed(const char* keymap)
{
    if (!StHashMap_contains(__st_keymaps__, keymap)) return false;

    StList* keys = StHashMap_get(__st_keymaps__, keymap);
    for (nint i = 0; i < keys->length; i++) {
        if (StInput_is_key_not_pressed((StKey)keys->items[i])) return true;
    }
    return false;
}

File: modules/platform/input.go
package platform

import rl "github.com/gen2brain/raylib-go/raylib"

/*bool StInput_is_keymap_just_pressed(const char* keymap)
{
    if (!StHashMap_contains(__st_keymaps__, keymap)) return false;

    StList* keys = StHashMap_get(__st_keymaps__, keymap);
    for (nint i = 0; i < keys->length; i++) {
        if (StInput_is_key_just_pressed((StKey)keys->items[i])) return true;
    }
    return false;
}*/

var keymapma map[string][]Key

func init() {
	keymapma = make(map[string][]Key)
}

func AddKeymap(keymap string, key Key) {
	keymapma[keymap] = append(keymapma[keymap], key)
}

func IsKeyJustPressed(key Key) bool {
	return rl.IsKeyPressed(int32(key))
}

func IsKeyJustReleased(key Key) bool {
	return rl.IsKeyReleased(int32(key))
}

func IsKeyJustHeld(key Key) bool {
	return rl.IsKeyDown(int32(key))
}

func IsKeyNotPressed(key Key) bool {
	return rl.IsKeyUp(int32(key))
}

func IsMouseButtonJustPressed(btn MouseButton) bool {
	return rl.IsMouseButtonPressed(rl.MouseButton(btn))
}

func IsMouseButtonJustReleased(btn MouseButton) bool {
	return rl.IsMouseButtonReleased(rl.MouseButton(btn))
}

func IsMouseButtonJustHeld(btn MouseButton) bool {
	return rl.IsMouseButtonDown(rl.MouseButton(btn))
}

func IsMouseButtonNotPressed(btn MouseButton) bool {
	return rl.IsMouseButtonUp(rl.MouseButton(btn))
}

func IsKeymapJustPressed(keymap string) bool {
	keymapdeez, ok := keymapma[keymap]

	// i'm not ok :(
	if !ok {
		return false
	}

	for _, v := range keymapdeez {
		if (IsKeyJustPressed(v)) {
			return true
		}
	}
	return false
}

func IsKeymapJustReleased(keymap string) bool {
	keymapdeez, ok := keymapma[keymap]

	// i'm not ok :(
	if !ok {
		return false
	}

	for _, v := range keymapdeez {
		if (IsKeyJustReleased(v)) {
			return true
		}
	}
	return false
}

func IsKeymapJustHeld(keymap string) bool {
	keymapdeez, ok := keymapma[keymap]

	// i'm not ok :(
	if !ok {
		return false
	}

	for _, v := range keymapdeez {
		if (IsKeyJustHeld(v)) {
			return true
		}
	}
	return false
}

func IsKeymapNotPressed(keymap string) bool {
	keymapdeez, ok := keymapma[keymap]

	// i'm not ok :(
	if !ok {
		return false
	}

	for _, v := range keymapdeez {
		if (IsKeyNotPressed(v)) {
			return true
		}
	}
	return false
}